### PR TITLE
migration: Add multifd migration scenario

### PR DESCRIPTION
**Pull request by balamuruhans**
Add migration test scenarios to test the recent Multifd feature
with migration parameters configuring non default values.

Qemu commit: 14 commits from 35374cbdff
Signed-off-by: Balamuruhan S <bala24@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by balamuruhans** [comment]
depends on https://github.com/avocado-framework/avocado-vt/pull/1644

File: qemu/tests/cfg/migrate.cfg [hunk 1]
```diff
@@ -14,14 +14,22 @@
     # vmstate_check = yes
     variants:
         - @default:
+        - with_multifd:
+            only tcp,unix,fd
+            migrate_capabilities = "{'x-multifd': 'on'}"
```
Comment (vivianQizhu) [inline qemu/tests/cfg/migrate.cfg]: Not quit sure, but  here https://wiki.qemu.org/Features/Migration-Multiple-fds shows `x-multiple-fds`

Comment (balamuruhans) [reply qemu/tests/cfg/migrate.cfg]: @vivianQizhu Thank you for being careful, I think the wiki is outdated. I got chance to test the qemu upstream patch.

Comment (vivianQizhu) [reply qemu/tests/cfg/migrate.cfg]: @balamuruhans Okay, thanks confirming and sharing it.

**Comment by balamuruhans** [comment]
**Results:**
```
JOB ID     : 5869e45ce559b0ba90b462a4222d105692f47804
JOB LOG    : /home/bala/avocado-fvt-wrapper/results/job-2018-07-04T19.47-5869e45/job.log
 (1/6) guest_migration.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-qemu.migrate.tcp.with_filter_off.with_multifd:  PASS (70.25 s)
 (2/6) guest_migration.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-qemu.migrate.unix.with_filter_off.with_multifd:  PASS (25.32 s)
 (3/6) guest_migration.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-qemu.migrate.fd.with_filter_off.with_multifd:  ERROR (19.56 s)
 (4/6) guest_migration.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-qemu.migrate.with_set_speed.tcp.with_filter_off.with_multifd:  PASS (72.03 s)
 (5/6) guest_migration.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-qemu.migrate.with_set_speed.unix.with_filter_off.with_multifd:  PASS (25.44 s)
 (6/6) guest_migration.qcow2.virtio_scsi.smp2.virtio_net.Guest.HostOS.ppc64le.powerkvm-qemu.migrate.with_set_speed.fd.with_filter_off.with_multifd:  ERROR (19.44 s)
RESULTS    : PASS 4 | ERROR 2 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 318.85 s
JOB HTML   : /home/bala/avocado-fvt-wrapper/results/job-2018-07-04T19.47-5869e45/results.html
```
we are observing segmentation fault with multifd migration using unix sockets,

> 2018-07-04 19:49:57,288 client           L0605 INFO | [qemu output] /tmp/aexpect_VWF5cQGW/aexpect-ayhElA.sh: line 1: 108304 **Segmentation fault**      MALLOC_PERTURB_=1 /home/bala/qemu/build/ppc64-softmmu/qemu-system-ppc64 -S -name 'avocado-vt-vm1' -machine pseries -nodefaults -vga std -chardev socket,id=qmp_id_qmpmonitor1,path=/var/tmp/avocado_qgdNZg/monitor-qmpmonitor1-20180704-194913-WBkoYA2F,server,nowait -mon chardev=qmp_id_qmpmonitor1,mode=control -chardev socket,id=qmp_id_catch_monitor,path=/var/tmp/avocado_qgdNZg/monitor-catch_monitor-20180704-194913-WBkoYA2F,server,nowait -mon chardev=qmp_id_catch_monitor,mode=control -chardev socket,id=serial_id_serial0,path=/var/tmp/avocado_qgdNZg/serial-serial0-20180704-194913-WBkoYA2F,server,nowait -device spapr-vty,reg=0x30000000,chardev=serial_id_serial0 -device nec-usb-xhci,id=usb1,bus=pci.0,addr=0x3 -device virtio-scsi-pci,id=virtio_scsi_pci0,bus=pci.0,addr=0x4 -drive id=drive_image1,if=none,format=qcow2,file=/home/bala/avocado-fvt-wrapper/data/avocado-vt/images/hostos-ppc64le.qcow2 -device scsi-hd,id=image1,drive=drive_image1 -device virtio-net-pci,mac=9a:79:7a:7b:7c:7d,id=idjGxD1g,vectors=4,netdev=idtt46a5,bus=pci.0,addr=0x5 -netdev tap,id=idtt46a5,fd=12 -m 1024 -smp 8,cores=1,threads=1,sockets=8 -device usb-tablet,id=usb-tablet1,bus=usb1.0,port=1 -vnc :0 -rtc base=utc,clock=host -boot menu=off,strict=off,order=cdn,once=c -enable-kvm -incoming unix:/var/tmp/avocado_qgdNZg/migration-unix-20180704-194913-WBkoYA2F

File: qemu/tests/migration.py [hunk 2]
```diff
@@ -210,6 +211,14 @@ def check_dma():
 
             capabilities = ast.literal_eval(params.get("migrate_capabilities", "{}"))
             inner_funcs = ast.literal_eval(params.get("migrate_inner_funcs", "[]"))
+            mig_parameters = ast.literal_eval(params.get("migrate_parameters", "{}"))
+            target_mig_parameters = params.get("target_migrate_parameters", "{}")
+            target_mig_parameters = ast.literal_eval(target_mig_parameters)
```
Comment (sathnaga) [inline qemu/tests/migration.py]: can be a single statement?, soft wrapped?

Comment (balamuruhans) [reply qemu/tests/migration.py]: It exceeded 80 char and the readability was not good, so I had split it in 2 lines

Comment (sathnaga) [reply qemu/tests/migration.py]: okies..

**Pull request merged**